src/core/server/http/router/request.ts

```diff
@@ -114,6 +114,11 @@ export class KibanaRequest<
    * This property will contain a `filtered` copy of request headers.
    */
   public readonly headers: Headers;
+  /**
+   * Whether or not the request is a "system API" call rather than an application-level call.
+   * Can be set on the client using the `asSystemApi` option.
+   */
+  public readonly isSystemApi: boolean;
```

Comment:
is there a client counterpart? This https://github.com/elastic/kibana/pull/53734/files#diff-98cc4b5fe935f0628aacb3c4918eb72bR43 is used here https://github.com/elastic/kibana/blob/c36f576998a7967173f3bcaea61af26753a551f5/x-pack/legacy/plugins/security/public/hacks/on_session_timeout.js#L19

Reply:
Ahh didn't catch that one. This will require a bit more refactoring to add a wrapper interface around Request.

Reply:
One decision that needs to be made here: should interceptors be able to override the asSystemApi flag?

Right now, we generate the full `Request` object before passing to interceptors, which would allow interceptors to remove the `kbn-system-api` header. Another issue with this implementation is that it exposes the "private" details of `kbn-version` and `kbn-system-api`, which probably shouldn't be visible to plugins at all.

What we probably should do is make the interceptor interface receive a HttpFetchOptions object and allow the interceptor to return a subset of this interface (`Partial<HttpFetchOptions>`) to override any options. We shouldn't generate the actual Request instance until after all interceptors have run. This would allow us to hide any private modifications that core does to the request, such as adding system headers.

I think for now, I will only change the arguments to interceptors to accept a IHttpRequest object:

```
interface IHttpRequest {
  request: Request,
  isSystemApi: boolean
}
```

But only allow interceptors to return a `Request` so there's only one mechanism for modifying the request. In a future PR, we should change this to  `HttpFetchOptions` / `Partial<HttpFetchOptions>` as mentioned above.

Reply:
I ended up implementing this change now, so now request interceptors modify the `HttpFetchOptions` rather than the raw Request. The `HttpResponse` interface now exposes the `fetchOptions` used to generate the Request which includes the `asSystemApi` option.